core/src/banking_stage.rs

```diff
@@ -378,13 +383,257 @@ pub enum ForwardOption {
+                        })
+                        .collect::<MinMaxHeap<_>>();
+
+                buffered_packet_batches.packet_priority_queue = retryable_packets;
+
+                // Assert unprocessed queue is still consistent
+                assert_eq!(
+                    buffered_packet_batches.packet_priority_queue.len(),
+                    buffered_packet_batches.message_hash_to_transaction.len()
+                );
+            }
+            Self::VoteStorage(latest_unprocessed_votes, VoteSource::Tpu) => {
+                // Based on the stake distribution present in the supplied bank, drain the unprocessed votes
+                // from each validator using a weighted random ordering. Votes from validators with
+                // 0 stake are ignored.
+                // TODO: Add proper error handling in case bank is missing?
+                let bank = bank.unwrap();
+
+                let latest_votes_per_pubkey = latest_unprocessed_votes
+                    .latest_votes_per_pubkey
+                    .read()
+                    .unwrap();
+
+                let retryable_votes = weighted_random_order_by_stake(&bank)
+                    .filter_map(|pubkey| {
+                        if let Some(lock) = latest_votes_per_pubkey.get(&pubkey) {
+                            if let Ok(latest_vote) = lock.write() {
+                                if let Ok(mut latest_vote) = latest_vote.try_borrow_mut() {
+                                    if !latest_vote.is_processed() {
+                                        let packet = latest_vote.clone();
+                                        latest_vote.clear();
+                                        latest_unprocessed_votes
+                                            .size
+                                            .fetch_sub(1, Ordering::AcqRel);
+                                        return Some(packet);
+                                    }
+                                }
+                            }
+                        }
+                        None
+                    })
+                    .chunks(batch_size)
+                    .into_iter()
+                    .flat_map(|vote_packets| {
+                        let vote_packets = vote_packets.into_iter().collect_vec();
+                        let packets_to_process = vote_packets
+                            .iter()
+                            .map(&DeserializedVotePacket::get_vote_packet)
+                            .collect_vec();
+                        if let Some(retryable_vote_indices) =
+                            processing_function(&packets_to_process)
+                        {
+                            retryable_vote_indices
+                                .iter()
+                                .map(|i| vote_packets[*i].clone())
+                                .collect_vec()
+                        } else {
+                            vote_packets
```

Comment:
A unit test making sure this function properly returns the retryable indices for the vote transaction storage vs regular transaction storage would be helpful to make sure this indexing stuff doesn't break